### ISSUE TITLE
ui: Improve Optional validation popup

### ISSUE THREAD

**Issue by Monviech** [opnsense/core#7611]
**Important notices**

Before you add a new report, we ask you kindly to acknowledge the following:

- [x] I have read the contributing guide lines at https://github.com/opnsense/core/blob/master/CONTRIBUTING.md
- [x] I am convinced that my issue is new after having checked both open and closed issues at https://github.com/opnsense/core/issues?q=is%3Aissue

**Is your feature request related to a problem? Please describe.**

This optional validation dialog looks a little outdated. I don't like that it shows a "bug" as a button, even though it is not a bug, it is a failed validation.

https://github.com/opnsense/core/blob/c0167704c3547ff11e63d523d4805d9fca401c0b/src/opnsense/www/js/opnsense_ui.js#L76

**Describe the solution you like**

I would like if the popup dialog would be nicer and less alarming in some way, showing right away what the problem is in a more user friendly way.

**Describe alternatives you considered**

It's mostly a cosmetic issue, so leaving it as it is.

**Additional context**

I'm using this dialog in the os-caddy plugin, since it shows otherwise hidden validations when the full model is validated. When it happens (which is quite often with the amount of validation), it is quite unexpected and it looks like a bug to the user (in my opinion).

**Comment by fichtner** [comment]
A bit more context in https://github.com/opnsense/core/issues/6898 and older tickets. Getting rid of the dialog at least for base components is alright but it may not be feasible in all use cases.

**Comment by Monviech** [comment]
@fichtner I don't want to get rid of that dialog. I just want it to look nicer and less alarming.

I need that dialog, it is important.

**Comment by fichtner** [comment]
Hmm, I don’t think it’s needed for the average use case. Browser network debug also is better in many cases. I guess it depends… for special cases special solutions can be much better and offer contained code that doesn’t break elsewhere or taints it with unused hooks… such as this dialog boolean enable that is very hard to remove from the respective JavaScript function. 😉

**Comment by Monviech** [comment]
I don't understand, I just want to improve the UX here.

Here's an example where I made it more clear that its a warning, and not a bug. And that it can be clicked to be expanded (it also has a tooltip)

![screen](https://github.com/opnsense/core/assets/79600909/4dc62832-2e1a-46e4-b673-55713a30faaa)

```
errorMessage.append('Please correct validation errors in form.<br/>');
errorMessage.append('<span>Click to toggle details: </span><i class="fa fa-exclamation-triangle" aria-hidden="true" style="cursor: pointer; color: #f0ad4e; margin-left: 5px;" title="Click to toggle details" data-toggle="collapse" ' +
                    'data-target="#' + detailsid + '" aria-expanded="false" aria-controls="' + detailsid + '"></i>');
errorMessage.append('<div class="collapse" id="' + detailsid + '"><hr/><pre></pre></div>');
```

If something like this is acceptable, I make a PR. If not, I can close this issue.

**Comment by fichtner** [comment]
> I don't understand, I just want to improve the UX here.

I get it. I just want to get rid of it instead. ;)

The concept of an "input validation" popup that is not human readable is just for debugging INO. Improving it takes time. Improving it for just 2 use cases is questionable, but feel free to propose a small PR.

I just think it's worth investing the time getting rid of it everywhere which would be higher value, longer term goal.


Cheers,
Franco

**Comment by Monviech** [comment]
If there is a better longer term goal I want to go there.

But, how can we get rid of it everywhere. For example here:

https://github.com/opnsense/plugins/blob/15c20bde5cd99fd094b85cf0fc13ad308743499c/www/caddy/src/opnsense/mvc/app/models/OPNsense/Caddy/Caddy.php#L221

The validations trigger on two pages, the `general.volt` and the `reverse_proxy.volt`, while the user only sees the `general.volt` view. And even when the user would go into the `reverse_proxy.volt`, they could not see any validation errors, since they would be hidden inside the forms of each item.

Now, if there would be no option for a popup dialog that also shows the validation error of "out of view scope" items, the form could never be saved and there would be no feedback.

**Comment by fichtner** [comment]
> The validations trigger on two pages

Ok, but how does the user know he/she should go to the other page? For tabs I added a click/focus to bring the validation error into user view. For Unbound and DNSmasq it also cross-checks ports just telling the user that the port us currently used by another service: "Unbound DNS is currently using this port." no popup but all relevant information.


Cheers,
Franco

**Comment by Monviech** [comment]
Oh... I think I understand. It would be better to append the validation error to an item in the form the user sees at the moment, additionally.

Then there is no need for a popup and there can be clear instructions.

Maybe I thought too complicated here.

**Comment by fichtner** [comment]
Yes. There are rare cases where none of the elements currently visible block a save, but that's more of a general data integrity issue then. Triggering the dialog in these cases might be feasible, but as said that's a different approach for it vs. how it's currently being used. it would require processing the list of validation errors and removing the ones that could be pinned to the page only showing the ones not on the page... but if you have one on the page and one elsewhere showing both is also a bit weird as you lose focus of the page one since the popup tells you about something else entirely.
